[clang-doc] add throws comments to comment template

evelez7 opened the pull request:
Serialize throw Doxygen comments for exceptions. Accepts both \throw and
\throws.

evelez7 committed:
[clang-doc] add throws comments to comment template

Serialize throw Doxygen comments for exceptions. Accepts both \throw and
\throws.

evelez7 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150649?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#150649** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150649?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150649?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#150648** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150648?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#150647** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#150571** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150571?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#150570** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/150570?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

ilovepi commented on clang-tools-extra/clang-doc/JSONGenerator.cpp:
```diff
@@ -126,6 +138,15 @@ static Object serializeComment(const CommentInfo &I, Object &Description) {
     auto TextCommentsArray = extractTextComments(CARef.front().getAsObject());
     if (I.Name == "brief")
       insertComment(Description, TextCommentsArray, "BriefComments");
+    else if (I.Name == "return")
+      insertComment(Description, TextCommentsArray, "ReturnComments");
```
I thought the return bits had their own patch? was that missing? split wrong?

ilovepi replied:
Oh, I see, there's a conflict. disregard.

evelez7 commented:
Think I lost track of this during GSoC submission. Should still be valid.

ilovepi reviewed (approved):
LGTM. It'd be nice if that HTML output looked sane, but if the tests still pass, its probably fine.

ilovepi commented on clang-tools-extra/test/clang-doc/basic-project.mustache.test:
```diff
@@ -384,6 +384,10 @@ HTML-CALC:                    </div>
+HTML-CALC:                    <h3>Throws</h3>
+HTML-CALC:                    <div>
+HTML-CALC:                        <b></b>     <div>
+HTML-CALC:                        <p>if b is zero.</p>
```
hmm, can we see what the new output is like? I just realized that extra tag wasn't a closing `</div>`.

ilovepi replied:
Looking at the doc string, I think there's missing data

ilovepi replied:
`std::invalid_argument` is missing from the output. I think we need to run down what's breaking

evelez7 commented:
The extra `div` isn't just present in the throws comments, it's present in the param comments as well. Here's an updated link for the output of this PR, and this class has param and throws comments: https://erickvelez.com/clang-doc-mustache-output/pr150649/GlobalNamespace/_ZTV10Calculator.html

What these two have in common is that they recursively call the comment partial to render the text comments.

https://github.com/llvm/llvm-project/blob/6429549a724003c0837f5d8a1ab911fc916f63eb/clang-tools-extra/clang-doc/assets/comment-template.mustache#L31-L31

There's no `<div>` on this line, so I'm not sure how one is appearing even if there is a missing placeholder data.

evelez7 commented:
I found the issue with the `<div>`. It's actually behaving correctly according to our implementation. The partial begins rendering on the line that it's called, so when the partial is called and matches with `{{#TextComment}}` on comment-template.mustache:77

https://github.com/llvm/llvm-project/blob/321a97e0f7c2536113bd0a901a8fd614e0a3faba/clang-tools-extra/clang-doc/assets/comment-template.mustache#L69

the `<div>` is added on the first line, then the rest of the partial is rendered on newlines. So, I was misusing it thinking that the text would be rendered on the same line. Fixed that. That means I should be able to fix the other weird whitespace stuff.

Also yes, the wrong placeholder was being named for the exception. Updated the link to verify: https://erickvelez.com/clang-doc-mustache-output/pr150649/GlobalNamespace/_ZTV10Calculator.html

ilovepi reviewed (approved):
Thanks for running that down. This looks great now.

evelez7 commented:
### Merge activity

* **Nov 15, 6:26 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/llvm/llvm-project/150649).
* **Nov 15, 6:28 PM UTC**: @evelez7 merged this pull request with [Graphite](https://app.graphite.com/github/pr/llvm/llvm-project/150649).
